src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,157 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_SIZE = 10;
+    private int size;
+    private T[] array;
+
+    public ArrayList() {
+        array = (T[]) new Object[DEFAULT_SIZE];
+    }
+
     @Override
     public void add(T value) {
-
+        checkSize();
+        array[size - 1] = value;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index can't be less than zero");
+        } else if (index < size) {
+            size++;
+            T[] tempArray = (T[]) new Object[checkLength()];
+            for (int i = 0; i < index; i++) {
+                tempArray[i] = array[i];
+            }
+            if (index < size) {
+                for (int i = index + 1; i < tempArray.length; i++) {
+                    tempArray[i] = array[i - 1];
+                }
+            }
+            tempArray[index] = value;
+            array = tempArray;
+        } else if (index == size) {
+            add(value);
+        } else {
+            throw new ArrayListIndexOutOfBoundsException("Invalid index");
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        T[] tempArray = (T[]) new Object[list.size()];
+        int index = 0;
+        while (index < list.size()) {
+            tempArray[index] = list.get(index);
+            index++;
+        }
+        resize(tempArray);
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return array[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkIndex(index);
+        array[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        T removedValue = array[index];
+        array = getArrayAfterRemove(index);
+        return removedValue;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        Integer indexToRemove = null;
+        for (int i = 0; i < array.length; i++) {
+            if (array[i] == null && element == null || array[i] != null
+                    && array[i].equals(element)) {
+                indexToRemove = i;
+                break;
+            }
+        }
+        if (indexToRemove == null) {
+            throw new NoSuchElementException("No such element in array");
+        }
+        array = getArrayAfterRemove(indexToRemove);
+        return element;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void checkSize() {
+        if (size >= DEFAULT_SIZE) {
+            resize();
```

Comment:
why do you make resize if  size >= DEFAULT_SIZE ?

Reply:
Because if size == DEFAULT_SIZE (10) I need increase size to +1 (11)  and also array.length otherwise I get - java.lang.ArrayIndexOutOfBoundsException

Reply:
![image](https://github.com/mate-academy/jv-array-list/assets/20326858/c11299fe-fb32-477a-a317-142a3bd847e3)